test/extensions/filters/network/ssh/service_userauth_test.cc

```diff
@@ -0,0 +1,746 @@
+  EXPECT_CALL(*transport_, initUpstream(_))
+    .WillOnce(SaveArg<0>(&state));
+
+  auto r = service_->handleMessage(std::move(msg));
+  ASSERT_OK(r);
+
+  ASSERT_THAT(*state->allow_response, Envoy::ProtoEq(*allow));
+  ASSERT_EQ(42, state->stream_id);
+  ASSERT_EQ(*state->downstream_ext_info, ext_info);
+  ASSERT_EQ(ChannelMode::Normal, state->channel_mode);
+}
+
+TEST_F(DownstreamUserAuthServiceTest, HandleMessageServerAllowInternal) {
+  auto msg = std::make_unique<pomerium::extensions::ssh::ServerMessage>();
+  auto allow = msg->mutable_auth_response()->mutable_allow();
+  auto internal = allow->mutable_internal();
+  ProtobufWkt::Value v;
+  v.set_string_value("example-metadata-value");
+  ProtobufWkt::Struct metadata_struct{};
+  (*metadata_struct.mutable_fields())["example-metadata-key"] = v;
+  auto filter_metadata = *internal->mutable_set_metadata()->mutable_filter_metadata();
+  filter_metadata["example-filter-name"] = metadata_struct;
+
+  EXPECT_CALL(*transport_, streamId())
+    .WillOnce(Return(42));
+  wire::ExtInfoMsg ext_info;
+  wire::test::populateFields(ext_info);
+  EXPECT_CALL(*transport_, peerExtInfo())
+    .WillOnce(Return(ext_info));
+
+  AuthStateSharedPtr state;
+  EXPECT_CALL(*transport_, initUpstream(_))
+    .WillOnce(SaveArg<0>(&state));
+
+  auto r = service_->handleMessage(std::move(msg));
+  ASSERT_OK(r);
+
+  ASSERT_THAT(*state->allow_response, Envoy::ProtoEq(*allow));
+  ASSERT_EQ(42, state->stream_id);
+  ASSERT_EQ(*state->downstream_ext_info, ext_info);
+  ASSERT_EQ(ChannelMode::Hijacked, state->channel_mode);
+}
+
+TEST_F(DownstreamUserAuthServiceTest, HandleMessageServerAllowUnsupportedTarget) {
+  auto msg = std::make_unique<pomerium::extensions::ssh::ServerMessage>();
+  msg->mutable_auth_response()->mutable_allow();
+  EXPECT_CALL(*transport_, streamId())
+    .WillOnce(Return(42));
+  EXPECT_CALL(*transport_, peerExtInfo())
+    .WillOnce(Return(wire::ExtInfoMsg{}));
+  auto r = service_->handleMessage(std::move(msg));
+  ASSERT_EQ(r.code(), absl::StatusCode::kInternal);
+  ASSERT_EQ(r.message(), "invalid target");
+}
+
+TEST_F(DownstreamUserAuthServiceTest, HandleMessageServerDeny) {
+  auto server_msg = std::make_unique<pomerium::extensions::ssh::ServerMessage>();
+  auto deny = server_msg->mutable_auth_response()->mutable_deny();
+  deny->set_partial(true);
+  deny->add_methods("publickey");
+  deny->add_methods("keyboard-interactive");
+
+  wire::Message ssh_msg;
+  EXPECT_CALL(*transport_, sendMessageToConnection(_))
+    .WillOnce(DoAll(SaveArg<0>(&ssh_msg),
+                Return(absl::UnknownError("sentinel"))));
+
+  auto r = service_->handleMessage(std::move(server_msg));
+  ASSERT_EQ(r, absl::UnknownError("sentinel"));
+
+  ASSERT_EQ(wire::SshMessageType::UserAuthFailure, ssh_msg.msg_type());
+  auto failure_msg = ssh_msg.message.get<wire::UserAuthFailureMsg>();
+  ASSERT_TRUE(failure_msg.partial);
+  ASSERT_EQ((string_list{"publickey", "keyboard-interactive"}), failure_msg.methods);
```

Comment:
This works, but it tends to be less error-prone to put the assertions inside the EXPECT_CALL matcher, rather than capturing the arg and asserting on it afterwards. This could be re-written as:
```cc
  EXPECT_CALL(*transport_, sendMessageToConnection(
                             MSG(wire::UserAuthFailureMsg,
                                 AllOf(FIELD_EQ(methods, string_list{"publickey", "keyboard-interactive"}),
                                       FIELD_EQ(partial, true)))))
    .WillOnce(Return(absl::UnknownError("sentinel")));

  auto r = service_->handleMessage(std::move(server_msg));
  ASSERT_EQ(r, absl::UnknownError("sentinel"));
```